File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.py
from collections import deque
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root is None:
            return None
        result=[]
        queue=deque()
        queue.append(root)
        while(queue):
            currentLength=len(queue)
            currentLevel=[]
            for i in range(currentLength):
                currentNode=queue.popleft()
                if(currentLength-i!=1):
                    nextNode=queue[0]
                    nextChecker=False
                else:
                    nextNode=None
                    nextChecker=True
                currentNode.next=nextNode
                currentLevel.append(currentNode.val)
                if nextChecker:
                    currentLevel.append('#')
                if currentNode.left:
                    queue.append(currentNode.left)
                if currentNode.right:
                    queue.append(currentNode.right)
            result+=currentLevel
        return root
                
        

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root==null){
            return root;
        }
        List<Object> result=new ArrayList<>();
        Deque<Node> queue=new ArrayDeque<>();
        queue.offer(root);
        while(!queue.isEmpty()){
            int currentQueueLength=queue.size();
            List<Object> currentQueueLevel=new ArrayList<>();
            for(int i=0;i<currentQueueLength;i++){
                Node currentNode=queue.poll();
                Node nextNode;
                boolean nullChecker;
                if(currentQueueLength-i==1){//edge case
                    nextNode=null;
                    nullChecker=true;
                }else{
                    nextNode=queue.peekFirst();
                    nullChecker=false;
                }
                if(currentNode.left!=null){
                    queue.offer(currentNode.left);
                }
                if(currentNode.right!=null){
                    queue.offer(currentNode.right);
                }
                currentNode.next=nextNode;
                currentQueueLevel.add(currentNode.val);
                if(nullChecker){
                    currentQueueLevel.add('#');
                }
            }
            result.addAll(currentQueueLevel);
        }
        
        return root;
        
        
    }
}